vm_api/src/trace.rs

```diff
@@ -157,18 +157,19 @@ impl ExpectInvocation {
     }
 
     pub fn fmt_invocs(&self, invocs: &[InvocationTrace]) -> String {
-        invocs
-            .iter()
-            .enumerate()
-            .map(|(i, invoc)| format!("{}: [{}:{}],\n", i, invoc.to, invoc.method))
-            .collect()
+        invocs.iter().enumerate().fold(String::new(), |mut s, (i, invoc)| {
+            use std::fmt::Write;
+            let _ = writeln!(s, "{}: [{}:{}],", i, invoc.to, invoc.method);
```

Comment:
clippy did not like creating lots of small strings, this new version will build a single string.

Reply:
When Clippy is making you replace a map+collect with a fold and macros, I think it has gone rather out of its swim lane!

Reply:
Well, I mean, it's replacing one macro with another. It's point _is_ valid: calling `format!` repeatedly will allocate a lot of strings while repeatedly calling `write!` on the same string won't. 

In this case, it doesn't matter. But this could definitely have a performance impact if this were performance critical code.

Reply:
Yeah but it's a linter, not a performance coach. It's supposed (IMO) to make the code more readable, not faster.

Reply:
It's supposed to teach you when there may be a "better" way to do things. I do wish there were a more elegant way to write this particular code, but the suggestion is valid.

Reply:
Think of it this way: rust is big and complicated, you likely wouldn't learn half the language or about half the new features without something looking over your shoulder and telling you when there's a better way to do something.

Reply:
> I do wish there were a more elegant way to write this particular code

there is, if you resist the functional siren song

```rust
use std::fmt::Write;
let mut s = String::new();
for (i, invoc) in invocs.iter().enumerate() {
    let _ = write!(&mut s, "{}: [{}:{}],\n", i, invoc.to, invoc.method);
}
s
```

simple, clear, maintainable

Reply:
🧜‍♀️